src/lib/pipe_lib/replacable_strict_pipe.ml

```diff
@@ -0,0 +1,93 @@
+open Core_kernel
+open Async_kernel
+
+(* NOTE: Data flow:
+      long_live_writer
+   -> long_live_reader
+   -actor-> tmp_swapped_writer
+   -> tmp_swapped_reader (tracked out of this structure)
+*)
+
+type ('data_in_pipe, 'pipe_kind, 'write_return) t =
+  { name : string
+  ; long_live_writer :
+      ('data_in_pipe, 'pipe_kind, 'write_return) Strict_pipe.Writer.t
+  ; long_live_reader : 'data_in_pipe Strict_pipe.Reader.t
+  ; mutable should_terminate : bool
+  ; mutable reader_request :
+      (string * 'data_in_pipe Strict_pipe.Reader.t Ivar.t) option
+  ; mutable tmp_swapped_writer :
+      ( 'data_in_pipe
+      , Strict_pipe.synchronous
+      , unit Deferred.t )
+      Strict_pipe.Writer.t
+      option
+  }
+
+let rec run (t : _ t) =
+  let process_reader_request () =
+    let%map.Option tmp_pipe_name, response = t.reader_request in
+    let tmp_reader, new_tmp_writer =
+      Strict_pipe.create ~name:tmp_pipe_name Strict_pipe.Synchronous
+    in
+    Option.iter ~f:Strict_pipe.Writer.kill t.tmp_swapped_writer ;
+    t.tmp_swapped_writer <- Some new_tmp_writer ;
+    Ivar.fill response tmp_reader ;
+    t.reader_request <- None ;
+    `Worked
+  in
+  let process_write () =
+    match t.tmp_swapped_writer with
+    | None ->
+        Deferred.return None
+    | Some tmp_swapped_writer -> (
+        match%bind Strict_pipe.Reader.read t.long_live_reader with
+        | `Ok data ->
+            let%map () = Strict_pipe.Writer.write tmp_swapped_writer data in
```

Comment:
if old pipe reader hangs without doing the last read and is not closed either, the whole "replaceable" pipe is going to hang forever.

After making `reader_request` an ivar, it's diligent to have `Deferred.both` here, so that in case the old reader does neither read nor close, we will still be able to fuse data into the new reader.

Reply:
TODO: before fixing the issue, let's write a unit test that will timeout after 5s if nothing happens, describing the scenario of a reader being neither read from nor closed.

We'll need a bunch of other unit tests, like that every element written to pipe appears on one of the consecutive readers, in the correct order.